02-microarray/differential-expression_microarray_02_several-groups.Rmd

```diff
@@ -0,0 +1,522 @@
+  theme_classic() # This makes some aesthetic changes
+```
+
+Yes! These results make sense.
+The WNT samples have much higher expression of ENSG00000128683 than the other samples. 
+
+## Write results to file
+
+The results in `stats_df` will be saved to our `results/` directory. 
+
+```{r}
+readr::write_tsv(stats_df, file.path(
+  results_dir,
+  "GSE37418_limma_results.tsv" # Replace with a relevant output name
+))
+```
+
+## Make volcano plots
+
+We'll use the `ggplot2` to make a set of volcano plots.
+But first, we need to set up our data for plotting. 
+We will need the p values from the individual contrasts as well as the log fold changes. 
+
+We can obtain the contrast p values from the `contrasts_fit` object and make it a longer format that the `ggplot()` function will want for plotting. 
+
+```{r}
+# Let's extract the contrast p values for each and convert them to -log10()
+contrast_p_vals_df <- -log10(contrasts_fit$p.value) %>%
+  # Make this into a data frame
+  as.data.frame() %>%
+  # Store genes as their own column
+  tibble::rownames_to_column("Gene") %>%
+  # Make this into long format
+  tidyr::pivot_longer(dplyr::contains("vsOther"),
+    names_to = "contrast",
+    values_to = "neg_log10_p_val"
+  )
+```
+
+Now let's extract the log fold changes from `stats_df`.
+
+```{r}
+# Let's extract the fold changes from `stats_df`
+log_fc_df <- stats_df %>%
+  # We only want to keep the `Gene` column as well
+  dplyr::select("Gene", dplyr::contains("vsOther")) %>%
+  # Make this a longer format
+  tidyr::pivot_longer(dplyr::contains("vsOther"),
+    names_to = "contrast",
+    values_to = "logFoldChange"
+  )
+```
+
+We can perform an `inner_join()` of both these datasets using their `Gene` columns. 
+
+```{r}
+plot_df <- log_fc_df %>%
+  dplyr::inner_join(contrast_p_vals_df,
+    by = c("Gene", "contrast"),
+    suffix = c("_log_fc", "_p_val")
+  )
+```
+
+Let's print out a preview of `plot_df`.
+
+```{r}
+# Print out what this looks like
+head(plot_df)
+```
+Let's declare what we consider to be significant levels for fold change and for -log10 p values. 
+By saving this as its own variable, we only need to change these cutoffs in one place if we want to adjust later. 
+
+```{r}
+# This is equivalent to p value < 0.05
+p_val_cutoff <- 1.301
+
+# Absolute value cutoff for fold changes
+abs_fc_cutoff <- 5
+```
+
+Now we can use these cutoffs to make a new variable that declares which genes we consider significant. 
+We will use some logic with `dplyr::case_when()` to do this. 
+By default, tidyverse functions like `dplyr::case_when()` will assume an unquoted name is a column name, but we need to put `!!` so it knows that `abs_fc_cutoff` and `p_val_cutoff` are lone objects and not within the `plot_df` data frame. 
+
+```{r}
+plot_df <- plot_df %>%
+  dplyr::mutate(
+    signif_label = dplyr::case_when(
+      abs(logFoldChange) > !!abs_fc_cutoff & neg_log10_p_val > !!p_val_cutoff ~ "p-val and FC",
```

Comment:
I don't love that I had to show them tidyeval but I would love even less for them to change the cutoff in 6 places.

Reply:
You don't need it here. bare `abs_fc_cutoff` and `p_val_cutoff` works just fine.

Reply:
I thought it should. But when I ran it something else must have bee messed up then. Hm. Great! Will delete.